[New resource] IntuneTrustedRootCertificateIOS: Adds support for iOS trusted root certs

dannyKBjj opened the pull request:
#### Pull Request (PR) description
<!--
      DSCResource module IntuneTrustedRootCertificateIOS. Adds support for  #microsoft.graph.iosTrustedRootCertificate
-->

#### This Pull Request (PR) fixes the following issues
None

#### Task list

<!--
    To aid community reviewers in reviewing and merging your PR, please take the time to run
    through the below checklist and make sure your PR has everything updated as required.

    Change to [x] for each task in the task list that applies to your PR. For those task that
    don't apply to you PR, leave those as is.
-->

- [x] Added an entry to the change log under the Unreleased section of the file CHANGELOG.md.
      Entry should say what was changed and how that affects users (if applicable), and
      reference the issue being resolved (if applicable).
- [x] Resource parameter descriptions added/updated in the schema.mof.
- [x] Resource documentation added/updated in README.md.
- [x] Resource settings.json file contains all required permissions.
- [x] Examples appropriately added/updated.
- [x] Unit tests added/updated.
- [x] New/changed code adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

dannyKBjj committed:
Initial module. Adds support for iOS trusted root certs

dannyKBjj commented:
Will be adding Android equivalent module soon

ricmestre commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateIOS/MSFT_IntuneTrustedRootCertificateIOS.psm1:
```diff
@@ -0,0 +1,726 @@
+    $PSBoundParameters.Remove('TenantId') | Out-Null
+    $PSBoundParameters.Remove('CertificateThumbprint') | Out-Null
+    $PSBoundParameters.Remove('ManagedIdentity') | Out-Null
+    $PSBoundParameters.Remove('AccessTokens') | Out-Null
```
You can use *Remove-M365DSCAuthenticationParameter -BoundParameters $PSBoundParameters*, which you should also use in *Test-TargetResource* as well. This cmdlet removes the Verbose parameter too so if you use you may omit its individual removal further down.

ricmestre replied:
See https://github.com/microsoft/Microsoft365DSC/blob/Dev/Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppConfigurationDevicePolicy/MSFT_IntuneAppConfigurationDevicePolicy.psm1 for reference

dannyKBjj replied:
Hi,

I've made the changes, but not sure what you mean by "This cmdlet removes the Verbose parameter too so if you use you may omit its individual removal further down."

ricmestre replied:
Remove-M365DSCAuthenticationParameter removes all authentication parameters, plus Ensure and Verbose from whatever you give as input to *BoundParameters*, so you don't have to call *$CreateParameters.Remove('Verbose') | Out-Null* and *$UpdateParameters.Remove('Verbose') | Out-Null* like you're doing in your code.

ricmestre commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateIOS/MSFT_IntuneTrustedRootCertificateIOS.psm1:
```diff
@@ -0,0 +1,726 @@
+
+            $ValuesToCheck.Remove($key) | Out-Null
+        }
+    }
```
Why complicate things here? Compare-M365DSCComplexObject can compare an whole object, even if it's an array, you don't need to iterate it. It also can compare the number of objects of the source and target, you don't need to do it manually.

Again see https://github.com/microsoft/Microsoft365DSC/blob/5ec749efe0485874e832b85b5d9a77fb37a07e3a/Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppConfigurationDevicePolicy/MSFT_IntuneAppConfigurationDevicePolicy.psm1#L600 for reference, this chunk is used quite literally in all Intune resources in *Test-TargetResource*.

dannyKBjj replied:
OK, I'll make the edit. It isn't that way in MSFT_IntuneWifiConfigurationPolicyWindows10 either (which is the one I randomly chose to base mine on!)

ricmestre replied:
Not your fault but the Wifi resources are not a good example since they don't work correctly and have not been maintained for a long time.

dannyKBjj replied:
Haha, trust me to pick that then.

dannyKBjj replied:
done

dannyKBjj committed:
Update MSFT_IntuneTrustedRootCertificateIOS.psm1

ricmestre commented on Modules/Microsoft365DSC/Examples/Resources/IntuneTrustedRootCertificateIOS/Example/localhost.mof:
```diff
@@ -0,0 +1,39 @@
+                        Name="Example";
+

+                    };
```
Please remove this file

dannyKBjj replied:
done

ricmestre commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateIOS/MSFT_IntuneTrustedRootCertificateIOS.psm1:
```diff
@@ -0,0 +1,718 @@
+        #region resource generator code
+        if ($null -eq $getValue)
+        {
+            $getValue = Get-MgBetaDeviceManagementDeviceConfiguration -Filter "DisplayName eq '$Displayname'" -ErrorAction SilentlyContinue | Where-Object `
```
Please add -All switch here otherwise it won't traverse all pages even if you're filtering by display name.

ricmestre replied:
Also don't forget this one.

dannyKBjj replied:
done

dannyKBjj committed:
Deleted unwanted localhost.mof

dannyKBjj committed:
Update MSFT_IntuneTrustedRootCertificateIOS.psm1

dannyKBjj committed:
Merge branch 'Dev' into MSFT_IntuneTrustedRootCertificateIOS

ricmestre commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateIOS/MSFT_IntuneTrustedRootCertificateIOS.psm1:
```diff
@@ -0,0 +1,696 @@
+    $nullResult.Ensure = 'Absent'
+    try
+    {
+        if (-not [string]::IsNullOrWhiteSpace($id)){ $getValue = Get-MgBetaDeviceManagementDeviceConfiguration -DeviceConfigurationId $id -ErrorAction SilentlyContinue }
```
The style used here for brackets are always to have their own line, like you did everywhere else. I think I just spotted the same problem on the other PR you just raised.

```powershell
if (...)
{
    ...
}
```

dannyKBjj replied:
fixed in both

dannyKBjj committed:
Update MSFT_IntuneTrustedRootCertificateIOS.psm1

ricmestre commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateIOS/MSFT_IntuneTrustedRootCertificateIOS.schema.mof:
Did you save the schema file in UTF16? Could you please try with UTF8 instead? Otherwise it shows up in GH as binary file.

dannyKBjj replied:
Not on purpose, but yes it was UTF16, have resaved all files as UTF8 in both PRs

ricmestre commented:
OK let me test this on my dev tenant to check if there's any issues and I'll get back to you

dannyKBjj committed:
All files now UTF-8

ricmestre commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateIOS/MSFT_IntuneTrustedRootCertificateIOS.psm1:
```diff
@@ -0,0 +1,698 @@
+    try
+    {
+        if (-not [string]::IsNullOrWhiteSpace($id)){ 
+		$getValue = Get-MgBetaDeviceManagementDeviceConfiguration -DeviceConfigurationId $id -ErrorAction SilentlyContinue
```
Use 4 spaces instead of tab

dannyKBjj replied:
done

ricmestre commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateIOS/MSFT_IntuneTrustedRootCertificateIOS.psm1:
```diff
@@ -0,0 +1,698 @@
+    $nullResult.Ensure = 'Absent'
+    try
+    {
+        if (-not [string]::IsNullOrWhiteSpace($id)){ 
```
Still not quite right, the bracket { must be right below if aligned to the closing bracket }

ricmestre replied:
Just missing this change now, please also change it accordingly in your other PR including everything you changed on this one and then I'll review it as well.

dannyKBjj replied:
Done. All edits made to this should be in the other PR as well. Will use these as my template going forward. Thanks.

ricmestre commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateIOS/MSFT_IntuneTrustedRootCertificateIOS.psm1:
```diff
@@ -0,0 +1,698 @@
+        {
+            Update-DeviceConfigurationPolicyAssignment -DeviceConfigurationPolicyId $policy.id `
+                -Targets $assignmentsHash `
+                -Repository 'deviceAppManagement/mobileAppConfigurations'
```
This isn't correct, I had to figure out why the assignments were not working, this is because the correct repository to use here is *'deviceManagement/deviceConfigurations'*

dannyKBjj replied:
Oh I thought they were, I'll look into it.

dannyKBjj replied:
corrected. Thankks.

ricmestre commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateIOS/MSFT_IntuneTrustedRootCertificateIOS.psm1:
```diff
@@ -0,0 +1,698 @@
+        $assignmentsHash = ConvertTo-IntunePolicyAssignment -IncludeDeviceFilter:$true -Assignments $Assignments
+        Update-DeviceConfigurationPolicyAssignment -DeviceConfigurationPolicyId $currentInstance.id `
+            -Targets $assignmentsHash `
+            -Repository 'deviceAppManagement/mobileAppConfigurations'
```
Same as above, correct repository is *'deviceManagement/deviceConfigurations'*

dannyKBjj replied:
ok thanks, fixed in this and in the other PR. Sorry about that.

ricmestre commented:
@dannyKBjj Please check my latest comments and make sure you also correct them in your other PR

ricmestre commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateIOS/MSFT_IntuneTrustedRootCertificateIOS.psm1:
```diff
@@ -0,0 +1,698 @@
+            Write-Verbose -Message "No Intune Trusted Root Certificate Policy for iOS with Id {$id} was found"
+            return $nullResult
+        }
+
```
Add a line here with the below, otherwise if the blueprint contains a bogus Id or from a different tenant the verbose messages will pick up the Id from the blueprint which will be incorrect.

```powershell
$Id = $getValue.Id
```

dannyKBjj replied:
Sorry I don't think I understand.. I want it to pick up the bogus ID from the blueprint don't I? That code would only execute if $getValue.Id was null?

ricmestre replied:
Look at the actual placement of where I'm asking to place it, yes you first want to search the policy by its Id which if it's bogus/from another tenant you then search by Display Name, if you then find it this policy will actually have its own Id which is different from what it's in the blueprint so if you place *$Id = $getValue.Id* on line 113 the next verbose message will show the actual Id of this policy instead of what's in the blueprint which is wrong

dannyKBjj replied:
ah ok. Thanks.

dannyKBjj committed:
Replaced tabs with spaces as requested.

dannyKBjj committed:
Latest fixes

 $Id = $getValue.Id added on line 113 to make sure verbose message displays correct ID.
Now searching correct repository deviceManagement/deviceConfigurations for assignments.

dannyKBjj committed:
Update MSFT_IntuneTrustedRootCertificateIOS.psm1

dannyKBjj committed:
Merge branch 'Dev' into MSFT_IntuneTrustedRootCertificateIOS

dannyKBjj committed:
Corrected bracket alignment on line 94

ykuijs committed:
Merge branch 'Dev' into MSFT_IntuneTrustedRootCertificateIOS

ykuijs commented:
@ricmestre Can this PR be merged? Has it been fully reviewed?

ricmestre commented:
@ykuijs Hi, yes this one has been reviewed and successfully tested and can be merged.

Please allow me a couple of hours to check the other one for Android.
